### PR TITLE
[CE] Refactoring of the topic tree structures and a performance improvement

### PR DESCRIPTION
**Pull request by Remit**
**Tickets**

[Ticket 1](https://hivemq.kanbanize.com/ctrl_board/42/cards/9727/details/)
[Ticket 2](https://hivemq.kanbanize.com/ctrl_board/42/cards/10246/details/)

**Motivation**

Improves performance when dispatching PUBLISH packets to shared subscriptions. Refactors structures used by topic tree for better developer experience.

**Changes**

Refactored structure of the topic tree.

Performance improvement when dispatching publishes to shared subscriptions.

### CODE REVIEW

File: src/main/java/com/hivemq/metrics/HiveMQMetrics.java [hunk 1]
```diff
@@ -187,5 +187,13 @@ public class HiveMQMetrics {
     public static final HiveMQMetric<Counter> WILL_MESSAGE_PUBLISHED_COUNT_TOTAL =
             HiveMQMetric.valueOf("com.hivemq.messages.will.published.count.total", Counter.class);
 
+    /**
+     * represents a {@link Counter}, which measures the current count of shared subscriptions
+     *
+     * @since 2022.2
```
Comment (Remit) [inline src/main/java/com/hivemq/metrics/HiveMQMetrics.java:193]: is correct?

Comment (LBrandl) [reply src/main/java/com/hivemq/metrics/HiveMQMetrics.java:193]: We don't typically add new metics to the community edition. I'm not personally opposed to it, but I remember that it was once an explicit decision to have advanced metrics as an enterprise feature or something.

File: src/main/java/com/hivemq/mqtt/topic/SubscriberWithQoS.java [hunk 2]
```diff
@@ -27,15 +27,13 @@
 
 /**
  * This represents a subscriber (client ID) with a Quality of Service Level
- *
- * @author Dominik Obermaier
```
Comment (LBrandl) [inline src/main/java/com/hivemq/mqtt/topic/SubscriberWithQoS.java:31]: I'm also not sure if we wanted to remove the authors form the CE.

Comment (Remit) [reply src/main/java/com/hivemq/mqtt/topic/SubscriberWithQoS.java:31]: @SgtSilvio do you happen to know this?

File: src/main/java/com/hivemq/extensions/services/publish/PublishServiceImpl.java [hunk 3]
```diff
@@ -15,7 +15,11 @@
  */
 package com.hivemq.extensions.services.publish;
 
-import com.google.common.util.concurrent.*;
```
Comment (hlohse) [inline src/main/java/com/hivemq/extensions/services/publish/PublishServiceImpl.java:18]: Please revert

File: src/main/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImpl.java [hunk 4]
```diff
@@ -29,21 +29,37 @@
 import com.hivemq.extension.sdk.api.services.exception.InvalidTopicException;
 import com.hivemq.extension.sdk.api.services.exception.NoSuchClientIdException;
 import com.hivemq.extension.sdk.api.services.general.IterationCallback;
-import com.hivemq.extension.sdk.api.services.subscription.*;
```
Comment (hlohse) [inline src/main/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImpl.java:32]: Please revert

File: src/main/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImpl.java [hunk 5]
```diff
@@ -29,21 +29,37 @@
...
+import com.hivemq.extension.sdk.api.services.subscription.SubscriberForTopicResult;
+import com.hivemq.extension.sdk.api.services.subscription.SubscriberWithFilterResult;
+import com.hivemq.extension.sdk.api.services.subscription.SubscriptionStore;
+import com.hivemq.extension.sdk.api.services.subscription.SubscriptionType;
+import com.hivemq.extension.sdk.api.services.subscription.SubscriptionsForClientResult;
+import com.hivemq.extension.sdk.api.services.subscription.TopicSubscription;
 import com.hivemq.extensions.ListenableFutureConverter;
-import com.hivemq.extensions.iteration.*;
```
Comment (hlohse) [inline src/main/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImpl.java:34]: Please revert

File: src/main/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImpl.java [hunk 6]
```diff
@@ -29,21 +29,37 @@
...
 import com.hivemq.mqtt.message.subscribe.Topic;
 import com.hivemq.mqtt.topic.tree.LocalTopicTree;
 import com.hivemq.mqtt.topic.tree.SubscriptionTypeItemFilter;
+import com.hivemq.mqtt.topic.tree.TopicTreeImpl;
 import com.hivemq.persistence.clientsession.ClientSessionSubscriptionPersistence;
 import com.hivemq.persistence.clientsession.callback.SubscriptionResult;
 import com.hivemq.util.Topics;
 
 import javax.inject.Inject;
-import java.util.*;
```
Comment (hlohse) [inline src/main/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImpl.java:46]: Please revert

File: src/main/java/com/hivemq/persistence/clientsession/SharedSubscriptionServiceImpl.java [hunk 7]
```diff
@@ -33,7 +33,12 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-import static com.hivemq.configuration.service.InternalConfigurations.*;
```
Comment (hlohse) [inline src/main/java/com/hivemq/persistence/clientsession/SharedSubscriptionServiceImpl.java:36]: Please revert

File: src/test/java/com/hivemq/extensions/services/publish/PublishServiceImplTest.java [hunk 8]
```diff
@@ -50,7 +50,11 @@
 import java.util.concurrent.ExecutorService;
 
 import static org.junit.Assert.assertEquals;
-import static org.mockito.ArgumentMatchers.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/extensions/services/publish/PublishServiceImplTest.java:53]: Please revert

File: src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java [hunk 9]
```diff
@@ -32,7 +32,13 @@
 import com.hivemq.extension.sdk.api.services.subscription.SubscriptionStore;
 import com.hivemq.extension.sdk.api.services.subscription.SubscriptionsForClientResult;
 import com.hivemq.extension.sdk.api.services.subscription.TopicSubscription;
-import com.hivemq.extensions.iteration.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java:35]: Please revert

File: src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java [hunk 10]
```diff
@@ -47,17 +53,33 @@
 import org.mockito.MockitoAnnotations;
 import util.TestException;
 
-import java.util.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java:50]: Please revert

File: src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java [hunk 11]
```diff
@@ -47,17 +53,33 @@
...
-import java.util.concurrent.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java:51]: Please revert

File: src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java [hunk 12]
```diff
@@ -47,17 +53,33 @@
...
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Predicate;
 
-import static org.junit.Assert.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java:54]: Please revert

File: src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java [hunk 13]
```diff
@@ -47,17 +53,33 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
-import static org.mockito.Mockito.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/extensions/services/subscription/SubscriptionStoreImplTest.java:57]: Please revert

File: src/test/java/com/hivemq/mqtt/InternalPublishServiceImplTest.java [hunk 14]
```diff
@@ -43,14 +43,21 @@
 
 import static com.hivemq.mqtt.handler.publish.PublishReturnCode.FAILED;
 import static com.hivemq.mqtt.handler.publish.PublishReturnCode.NO_MATCHING_SUBSCRIBERS;
-import static org.junit.Assert.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/mqtt/InternalPublishServiceImplTest.java:46]: Please revert

File: src/test/java/com/hivemq/mqtt/InternalPublishServiceImplTest.java [hunk 15]
```diff
@@ -43,14 +43,21 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
-import static org.mockito.Mockito.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/mqtt/InternalPublishServiceImplTest.java:49]: Please revert

File: src/test/java/com/hivemq/mqtt/services/PublishPollServiceImplTest.java [hunk 16]
```diff
@@ -60,9 +60,17 @@
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static org.junit.Assert.assertEquals;
-import static org.mockito.ArgumentMatchers.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/mqtt/services/PublishPollServiceImplTest.java:63]: Please revert

File: src/test/java/com/hivemq/mqtt/services/PublishPollServiceImplTest.java [hunk 17]
```diff
@@ -60,9 +60,17 @@
...
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.anyLong;
-import static org.mockito.Mockito.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/mqtt/services/PublishPollServiceImplTest.java:65]: Please revert

File: src/test/java/com/hivemq/mqtt/topic/tree/TopicTreeStartupTest.java [hunk 18]
```diff
@@ -42,13 +42,12 @@
 import static org.hamcrest.CoreMatchers.hasItems;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
-import static org.mockito.ArgumentMatchers.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/mqtt/topic/tree/TopicTreeStartupTest.java:45]: Please revert

File: src/test/java/com/hivemq/persistence/clientqueue/ClientQueuePersistenceImplTest.java [hunk 19]
```diff
@@ -48,8 +48,16 @@
 
 import static com.hivemq.configuration.service.MqttConfigurationService.QueuedMessagesStrategy;
 import static org.junit.Assert.assertEquals;
-import static org.mockito.ArgumentMatchers.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/persistence/clientqueue/ClientQueuePersistenceImplTest.java:51]: Please revert the whole file

File: src/test/java/com/hivemq/persistence/clientsession/ClientSessionSubscriptionPersistenceImplTest.java [hunk 20]
```diff
@@ -43,9 +43,18 @@
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNull;
-import static org.mockito.ArgumentMatchers.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/persistence/clientsession/ClientSessionSubscriptionPersistenceImplTest.java:46]: Please revert the whole file

File: src/test/java/com/hivemq/persistence/clientsession/SharedSubscriptionServiceTest.java [hunk 21]
```diff
@@ -30,9 +30,15 @@
 import java.util.concurrent.ExecutionException;
 
 import static com.hivemq.persistence.clientsession.SharedSubscriptionServiceImpl.splitTopicAndGroup;
-import static org.junit.Assert.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/persistence/clientsession/SharedSubscriptionServiceTest.java:33]: Please revert the whole file

File: src/main/java/com/hivemq/mqtt/topic/tree/MatchingNodeSubscriptions.java [hunk 22]
```diff
@@ -0,0 +1,423 @@
...
+
+        if (nonSharedSubscriptions == null) {
+            return sharedSubscriptions;
+        }
+
+        return Stream.concat(sharedSubscriptions, nonSharedSubscriptions);
+    }
+
+    @ReadOnly
+    public @NotNull Set<SubscriberWithQoS> getAllSubscribers() {
```
Comment (hlohse) [inline src/main/java/com/hivemq/mqtt/topic/tree/MatchingNodeSubscriptions.java:218]: Seems to be unused

File: src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeImpl.java [hunk 23]
```diff
@@ -41,6 +40,8 @@
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.function.BiConsumer;
```
Comment (hlohse) [inline src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeImpl.java:42]: Seems to be unused now

File: src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeNode.java [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+                return previousValue;
+            }
+        } else {
+            children = new TopicTreeNode[]{node};
+        }
+
+        return node;
+    }
+
+    public @Nullable TopicTreeNode getChildNodeForTopicPart(final @NotNull String part) {
```
Comment (hlohse) [inline src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeNode.java:115]: Seems to be unused

File: src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeNode.java [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+    public @Nullable Map<String, TopicTreeNode> getChildrenMap() {
+        return childrenMap;
+    }
+
+    public @NotNull String getTopicPart() {
+        return topicPart;
+    }
+
+    @VisibleForTesting
+    public int getSegmentSubscriptionCount() { // Not used in production code
```
Comment (hlohse) [inline src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeNode.java:165]: Seems to be unused

File: src/test/java/com/hivemq/mqtt/topic/tree/TestGetSubscribersWithFilterFromTopicTreeImpl.java [hunk 26]
```diff
@@ -24,62 +23,56 @@
...
 import org.junit.Before;
 import org.junit.Test;
 
 import java.util.Random;
 import java.util.Set;
+import java.util.function.Predicate;
 
 import static org.hamcrest.CoreMatchers.hasItem;
 import static org.hamcrest.CoreMatchers.hasItems;
-import static org.junit.Assert.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/mqtt/topic/tree/TestGetSubscribersWithFilterFromTopicTreeImpl.java:36]: Please revert

File: src/test/java/com/hivemq/mqtt/topic/tree/TestTopicTreeImplEdgeCases.java [hunk 27]
```diff
@@ -32,35 +32,32 @@
 import java.util.Set;
 
 import static org.hamcrest.CoreMatchers.hasItem;
-import static org.junit.Assert.*;
```
Comment (hlohse) [inline src/test/java/com/hivemq/mqtt/topic/tree/TestTopicTreeImplEdgeCases.java:35]: Please revert

**Review by hlohse — CHANGES_REQUESTED** [review]
Looks good besides some import and unused methods nitpicking 👮‍♂️

**Comment by Remit** [comment]
> Looks good besides some import and unused methods nitpicking 👮‍♂️

Intellij found some default scheme and applied it  🤦 thanks for noticing!

**Review by hlohse — APPROVED** [review]

File: src/main/java/com/hivemq/metrics/HiveMQMetrics.java (hunk 1 above)
Comment (Remit) [reply src/main/java/com/hivemq/metrics/HiveMQMetrics.java:186]: removed along with one metric that was not used at all

File: src/main/java/com/hivemq/metrics/HiveMQMetrics.java [hunk 28]
```diff
@@ -187,5 +180,13 @@ public class HiveMQMetrics {
...
+     */
+    public static final HiveMQMetric<Counter> SHARED_SUBSCRIPTIONS_CURRENT =
```
Comment (LBrandl) [inline src/main/java/com/hivemq/metrics/HiveMQMetrics.java:188]: The metric is still here. It's now unused.

File: src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeImpl.java [hunk 29]
```diff
@@ -460,28 +397,33 @@ private Node getIndexForChildNode(final @NotNull String key, final @NotNull Node
      */
     private boolean removeRootWildcardSubscriber(final @NotNull String subscriber, final @Nullable String sharedName) {
         final ImmutableList.Builder<SubscriberWithQoS> foundSubscribers = ImmutableList.builder();
+        int sharedSubscriptionCount = 0;
```
Comment (LBrandl) [inline src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeImpl.java:400]: Counter isn't used anymore.

File: src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeImpl.java [hunk 30]
```diff
@@ -460,28 +397,33 @@ private Node getIndexForChildNode(final @NotNull String key, final @NotNull Node
...
         for (final SubscriberWithQoS rootWildcardSubscriber : rootWildcardSubscribers) {
             if (rootWildcardSubscriber.getSubscriber().equals(subscriber) &&
                     Objects.equals(rootWildcardSubscriber.getSharedName(), sharedName)) {
                 foundSubscribers.add(rootWildcardSubscriber);
+                if (rootWildcardSubscriber.isSharedSubscription()) {
```
Comment (LBrandl) [inline src/main/java/com/hivemq/mqtt/topic/tree/TopicTreeImpl.java:405]: If-block can be removed

File: src/main/java/com/hivemq/mqtt/topic/SubscriberWithQoS.java (hunk 2 above)
Comment (Remit) [reply src/main/java/com/hivemq/mqtt/topic/SubscriberWithQoS.java:31]: @LBrandl Silvio told me that the rule of not removing authors from APIs and removing them from impls applies to all of our projects, including CE

**Review by LBrandl — APPROVED** [review]

**Review by A-Imal — APPROVED** [review]

**Pull request merged**